### PR TITLE
Lower MatrixAggregateRowsByKey

### LINKED ISSUE #5184: lower MatrixAggregateRowsByKey

### PR DESCRIPTION
**Pull request by tpoterba**
Fixes #5184

### CODE REVIEW

**Comment by jigold** [comment]
I'll look at this tomorrow. Please remind me if I get sidetracked...

**Comment by tpoterba** [comment]
This depends on #5479 now

**Comment by tpoterba** [comment]
This is ready for review

File: hail/src/main/scala/is/hail/expr/ir/LowerMatrixIR.scala [hunk 1]
```diff
@@ -288,6 +288,20 @@ object LowerMatrixIR {
               irRange(0, 'global(lengths)(colIdx), 1).map(Symbol(genUID()) ~> 'row(entriesField)(colIdx)))))
         .mapGlobals('global.dropFields(lengths))
 
+    case MatrixAggregateRowsByKey(child, entryExpr, rowExpr) =>
+      lower(child)
+        .aggregateByKey(
+          aggLet(va = 'row) {
+            rowExpr.insertFields(entriesField -> irRange(0, 'global (colsField).len)
```
Comment (jigold) [inline hail/src/main/scala/is/hail/expr/ir/LowerMatrixIR.scala]: rowExpr is not used in here at all!

File: hail/src/main/scala/is/hail/expr/ir/ExtractAggregators.scala [hunk 2]
```diff
@@ -142,13 +142,32 @@ object ExtractAggregators {
                 ArrayRange(I32(0), ArrayLen(Ref(aUID, a.typ)), I32(1)),
                 iUID,
                 Let(
-                  name,
+                  elementName,
                   ArrayRef(Ref(aUID, a.typ), Ref(iUID, TInt32())),
-                  SeqOp(
+                  addLets(SeqOp(
```
Comment (jigold) [inline hail/src/main/scala/is/hail/expr/ir/ExtractAggregators.scala]: I'm having trouble understanding why the `addLets` got moved here.

File: hail/src/main/scala/is/hail/expr/ir/ExtractAggregators.scala [hunk 3]
```diff
@@ -142,13 +142,32 @@ object ExtractAggregators {
...
+          rUID,
+          GetTupleElement(result, i),
+          ArrayMap(
+            ArrayRange(
+              I32(0),
+              ArrayLen(Ref(rUID, rt)),
+              I32(1)),
+            indexName,
+            Let(
+              newRef.name,
```
Comment (jigold) [inline hail/src/main/scala/is/hail/expr/ir/ExtractAggregators.scala]: Is this mapping over each aggregator and that's what `indexName` corresponds to? Index in the aggregator array?

File: hail/python/hail/ir/ir.py [hunk 4]
```diff
@@ -919,32 +919,34 @@ def _compute_type(self, env, agg_env):
 
 
 class AggArrayPerElement(IR):
-    @typecheck_method(array=IR, name=str, agg_ir=IR, is_scan=bool)
-    def __init__(self, array, name, agg_ir, is_scan):
+    @typecheck_method(array=IR, element_name=str, index_name=str, agg_ir=IR, is_scan=bool)
+    def __init__(self, array, element_name, index_name, agg_ir, is_scan):
```
Comment (jigold) [inline hail/python/hail/ir/ir.py]: Can you describe what `index_name` represents just to make sure what I think it is (index into aggregator array) is correct?

File: hail/src/main/scala/is/hail/expr/ir/LowerMatrixIR.scala [hunk 5]
```diff
@@ -288,6 +288,20 @@ object LowerMatrixIR {
...
+              .aggElements('__element_idx, '__result_idx)(
+                  let(sa = 'global(colsField)('__result_idx)) {
+                    aggLet(sa = 'global (colsField)('__element_idx),
```
Comment (jigold) [inline hail/src/main/scala/is/hail/expr/ir/LowerMatrixIR.scala]: Why is `sa` defined twice?

File: hail/src/main/scala/is/hail/expr/ir/ExtractAggregators.scala (hunk 3 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/ExtractAggregators.scala]: no, I had to add this to support the middle line here:

```
              .aggElements('__element_idx, '__result_idx)(
                  let(sa = 'global(colsField)('__result_idx)) {
                    aggLet(sa = 'global (colsField)('__element_idx),
```

The post-aggregation operation needs to have `sa` in scope, so I need a way to expose a per-element context. The simplest thing was to bind the element index in the post-aggregation scope, and this is the emitted code to do that.

File: hail/src/main/scala/is/hail/expr/ir/ExtractAggregators.scala (hunk 2 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/ExtractAggregators.scala]: This was a bug - the lets added here are ones that mention the `elementName` binding, so they need to go inside the binding!

This fix was required to pass tests.

File: hail/src/main/scala/is/hail/expr/ir/LowerMatrixIR.scala (hunk 1 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/LowerMatrixIR.scala]: `thing.insertFields(bindings...)` generates a 

```


```

File: hail/src/main/scala/is/hail/expr/ir/LowerMatrixIR.scala (hunk 5 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/LowerMatrixIR.scala]: We need it once in the aggregation scope, and once in the post-aggregation scope

**Comment by danking** [comment]
@tpoterba should the review be dismissed? Can we close until further developments?

File: hail/python/hail/ir/ir.py (hunk 4 above)
Comment (tpoterba) [reply hail/python/hail/ir/ir.py]: (described in answer to the other comment)

File: hail/src/main/scala/is/hail/expr/ir/LowerMatrixIR.scala [hunk 6]
```diff
@@ -288,6 +288,20 @@ object LowerMatrixIR {
...
+                      g = 'row (entriesField)('__element_idx),
+                      sa = 'global (colsField)('__element_idx)) {
```
Comment (jigold) [inline hail/src/main/scala/is/hail/expr/ir/LowerMatrixIR.scala]: Fix dup.

**Pull request merged**